src/contracts/StakingChef.sol

```diff
@@ -0,0 +1,158 @@
+    uint256 private _totalAllocPoint;
+    uint256 private _lifetimeShareValue;
+
+    MetricToken private _metric;
+
+    constructor(address metricTokenAddress) {
+        _metric = MetricToken(metricTokenAddress);
+        setMetricPerBlock(4);
+    }
+
+    mapping(address => bool) public override addressExistence;
+    // modifier nonDuplicated(msg.sender);
+
+// --------------------------------------------------------------------- staking functions
+    function stakeMetric(
+        address newAddress,
+        uint256 metricAmount,
+        uint256 newStartDate
+    ) external nonDuplicated(newAddress) {
+    if (areRewardsActive() && _totalAllocPoint > 0) {
+            updateAccumulatedStakingRewards();
+        }
+        Staker memory stake = Staker({
+            stakeAddress: newAddress,
+            metricAmount: metricAmount, 
+            startDate: newStartDate,
+            rewardDebt: metricAmount.mul(_lifetimeShareValue).div(ACC_METRIC_PRECISION),
+            claimable: 0
+        });
+
+        _stakes.push(stake);
+        _totalAllocPoint = _totalAllocPoint.add(stake.metricAmount);
+        _metric.transfer(address(this), stake.metricAmount);
+        
+    }
+
+    function updateStaker(
+        address stakeAddress,
+        uint256 stakeIndex,
+        uint256 metricAmount 
+    ) public {
+        if (areRewardsActive() && _totalAllocPoint > 0) {
+            updateAccumulatedStakingRewards();
+        }
+        _totalAllocPoint = _totalAllocPoint.sub(_stakes[stakeIndex].metricAmount).add(metricAmount);
+        _stakes[stakeIndex].stakeAddress = stakeAddress;
+        _stakes[stakeIndex].metricAmount = metricAmount;
+    }
+
+    function removeStaker(uint256 stakeIndex) external {
+        require(stakeIndex < _stakes.length);
+        if (areRewardsActive() && _totalAllocPoint > 0) {
+            updateAccumulatedStakingRewards();
+        }
+        _totalAllocPoint = _totalAllocPoint.sub(_stakes[stakeIndex].metricAmount);
+
+        _stakes[stakeIndex] = _stakes[_stakes.length - 1];
+        _stakes.pop();
+    }
+
+    function stakeAdditionalMetric(
+        address stakeAddress,
+        uint256 stakeIndex,
+        uint256 metricAmount,
+        uint256 newStartDate
+    ) public {
+        harvest(stakeIndex);
+        uint256 principalMetric = _stakes[stakeIndex].metricAmount;
+        uint256 totalMetricStaked = metricAmount + principalMetric;
+
+        Staker memory stake = Staker({
+            stakeAddress: stakeAddress,
+            metricAmount: totalMetricStaked,
+            startDate: newStartDate,
+            rewardDebt: metricAmount.mul(_lifetimeShareValue).div(ACC_METRIC_PRECISION),
+            claimable: 0
+            });
+
+            _stakes.push(stake);
+            _totalAllocPoint = _totalAllocPoint.add(stake.metricAmount);
+            _metric.transfer(address(this), stake.metricAmount);
+    }
+
+    function updateAccumulatedStakingRewards() public {
```

Comment:
Let's work on adding this functionality to Chef - honestly I would like to get a deeper understanding of this functionality as it lives here and TopChef at the moment